chore(xunit): refactor xUnit reporter class

tymonx opened the pull request:
- Fixing/closing all related issues/PRs with JUnit XML tests report
- Based on [pytest internal JUnit XML plugin](https://github.com/pytest-dev/pytest/blob/9.0.2/src/_pytest/junitxml.py)
- Compatible with [Jenkins xUnit schema version 2](https://github.com/jenkinsci/xunit-plugin/blob/xunit-2.4.0/src/main/resources/org/jenkinsci/plugins/xunit/types/model/xsd/junit-10.xsd)
- Junit XML report file will wok with GitHub Actions, GitLab CI, Jenkins CI and other compatible CI environments
- Added documentation about generated JUnit XML `docs/source/junit.rst`
- New environment variable `COCOTB_RESULTS_ATTACHMENTS` that can be used to attach files in JUnit XML report file
- New environment variable `COCOTB_RESULTS_RELATIVE_TO`  that can be used to change all absolute paths into relative ones
- Created files (logs, waveforms, ...) can be attached to generated xUnit XML file
- Moved invalid XML element attributes to properties
- Added `attachment` properties to test cases
- Report attachments in `system-out`
- Report random seed in `system-err`
- Report exception in `failure` with message, type and traceback
- Group test cases in test suites based on name of Python module
- Removed `COCOTB_RESULT_TESTPACKAGE` environment variable, the `package` attribute is invalid in xUnit XML
- Removed `COCOTB_RESULT_TESTSUITE` environment variable, test cases are grouped in test suites based on name of Python module
- Unit tests for xUnit reporter class
- Hit `100%` code coverage for `src/cocotb/_xunit_reporter.py` file

Fixes #5185, fixes #4864, fixes #4171, fixes #4756, fixes #4165, fixes #3835, fixes #2513, closes #2488, closes #4854

<!--

Thanks for improving cocotb! Here are some points to make this as smooth as possible.
Not all of them may be applicable.

Most important: please explain *why* you are proposing this change.

* Make sure you have read https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md
* Extend or add a test under `tests/test_cases/`.
* Add documentation under `docs/source/`,
  docstrings in Python code, or Doxygen markup in C/C++ code.
  Use ``versionadded``/``versionchanged``/``deprecated``.
* Add a newsfragment - see `docs/source/newsfragments/README.rst`.
* Use `closes #XXXX` to auto-close the issue that this PR fixes (if such).

-->

tymonx committed:
chore(xunit): refactor XUnit reporter class

- Fix/close all related issues/PRs with JUnit XML tests report
- Based on pytest JUnit XML plugin
- Compatible with XUnit version 1.x (`xunit1`) and 2.x (`xunit2`)
- Automatically detect environment: GitHub Actions, GitLab CI, Jenkins
- New environment variable `COCOTB_ATTACHMENTS`
- Attach files (logs, waveforms, ...) to generated JUnit XML file
- Move invalid XML element attributes to properties
- Add `attachment` properties to test cases
- Report attachments in `system-out`
- Report random seed in `system-err`
- Report exception in `failure` with message, type and traceback
- Group test cases in test suites based on name of Python module
- Removed `COCOTB_RESULT_TESTPACKAGE` environment variable,
  the `package` attribute is invalid in XUnit
- Removed `COCOTB_RESULT_TESTSUITE` environment variable,
  test cases are grouped in test suites based on name of Python module
- Unit tests for XUnit reporter class
- Hit `100%` code coverage for `src/cocotb/_xunit_reporter.py` file

Fixes #5185, fixes #4864, fixes #4171, fixes #4756, fixes #3835,
fixes #2513, closes #2488, closes #4854

ktbarrett reviewed (commented):
Excellent! Thank you for taking this on.

I would also add "removal" newsfragments for removing the undocumented `COCOTB_RESULT_TESTSUITE` and `TESTPACKAGE` environment variables. Something like:

```
Removed support for undocumented ``COCOTB_RESULT_TESTSUITE`` environment variable.
```

tymonx commented on src/cocotb_tools/runner.py:
```diff
@@ -525,6 +525,17 @@ def test(
+        if waves_file:
+            attachments.append(self.test_dir / waves_file)
+
+        self.env["COCOTB_ATTACHMENTS"] = ",".join(map(str, attachments))
```
@ktbarrett Fixed in this commit e92f6073118ee409aecd2745e5dae86b506e9499. It is possible to override it using `extra_env` argument from the `Runner.test()` method or from global environment variable that will override it by the `_set_env_test()` private method that is invoked later.

tymonx commented on docs/source/newsfragments/5219.feature.1.rst:
```diff
@@ -0,0 +1 @@
+Tests reported in generated JUnit XML tests results file will be grouped into test suites per Python module.
```
@ktbarrett Fixed in this commit 3133db1d7daa0ce75edadb6c55b64b2652a9875c

imphil commented on docs/source/library_reference.rst:
```diff
@@ -195,11 +195,13 @@ Discovering Tests
 
+    .. versionchanged:: 2.1
+        Generated XML file is fully compatible with xUnit version 1.x (`xunit1`_) and 2.x (`xunit2`_).
+        The default xUnit format is `xunit2`_. The `xunit1`_ format is used in GitLab CI environments.
```
I'm confused -- the output is compatible with both schemas? How does it then matter what GitLab CI "uses" (as in, can consume?)

tymonx replied:
@imphil Bad wording. I will "refactor" it. Please refer to my comment https://github.com/cocotb/cocotb/pull/5219#issuecomment-3694255165

GitLab is happy with both versions but with v1 it can use `file` attribute to link source file with test in their tests report: [GitLab JUnit format specification](https://docs.gitlab.com/ci/testing/unit_test_reports/#junit-xml-format-specification). Where Jenkins is unhappy with v1.

cmarqu commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,398 @@
+            classname:   Path to module (using the ``.`` dot as separator) where test is located.
+            file:        Path to file with test.
+            line:        Line number of test in file.
+            time:        Real-time execution of test in seconds.
```
```suggestion
            time:        Wall clock execution time of test in seconds.
```

imphil reviewed (commented):
Thanks for your work on this feature! Some things I thought of when reading through this PR (not going into the implementation details yet).

As you said, naming is hard. Let's not make it worse for our users. What's implemented here is "the format understood by the Jenkins xunit plugin, either in version 1, or version 2"; they own the XML Schema. This format is then compatible with or equivalent to others (notably JUnit), but that's a secondary effect. And equally important, it has nothing to do with https://xunit.net/docs/format-xml-v2. So let's talk about the "Jenkins xUnit" (or something similar) in code and documentation.

I'd also appreciate a documentation page or something like that discoverable and targeting end users explaining (roughly) how the output format looks and what they can expect from it (that's clearly missing at the moment), including a sample XML snippet. Doesn't need to be a full format reference!

You referenced pytest a couple times. Which format is pytest implementing? Since we're in the Python world, it might make sense to compare what we're doing to pytest in the docs (briefly).

I'm not a fan of magic depending on the environment. It's rather surprising to users (especially verification engineers who take the ability to get identical reruns very seriously) to run the same test in a different environment and get different output. Maybe we can avoid that, or at least make it more discoverable.

- Why do we need the xunit1 format at all? Isn't v2 a subset of v1? I.e., if we only support v2, we loose a bit of information and can have an argument about backwards-compat with prior cocotb versions (which I'm willing to  have).
- If we can't get rid of the v1 format, then we at least need a way to let users configure what they want, e.g. an environment variable along the lines of `COCOTB_RESULTS_FILE_FORMAT=auto|jenkins_xunit1_xml|jenkins_xunit2_xml`.

Can you add a test that validates the resulting XML against the schema specified by Jenkins xunit?

Avoid names like "legacy". Today's "new" is tomorrow's "legacy". A unique name like "jenkins_xunit1_xml" or whatever it is will remain descriptive.

cmarqu commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,398 @@
+            attachments: List of attachments to add.
+            failure:     Fail test case.
+            error:       Error test case.
+            skipped:     Skip test case.
```
Can more than one of these be set?

tymonx replied:
@cmarqu It depends on XML parser and supported schema. In most cases:

- `failure`  can occur 0 .. N times
- `error` should occur only 0 or 1 time
- `skipped` should accur only 0 or 1 time

Any of them can occur in any combination. Test can thrown several exceptions (`failure`) and be marked as `skipped`. In the same time, it can be marked also as `error` if something went wrong with test runner, environment or something else not strictly related with test. Skipped should have higher precedence than `failure`/`error` :thinking: 

Good example could be sub-tests. Each sub-test can thrown an exception (`failure`). The main test of sub-tests can later thrown `Skip` exception or something similar on runtime. And test runner `teardown` could raise an internal error (`error`) after everything :melting_face:

Example: [pytest-subtests](https://pypi.org/project/pytest-subtests/). Sub-tests are not reported as separate test cases (at least for now).

tymonx commented:
@imphil Yes, I agree. I will try to rename them to `jenkins_xunit_v1|2`. And drop the `legacy` wording. It was heavily inspired by the pytest internal plugin [junitxml](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/junitxml.py).

> Why do we need the xunit1 format at all? Isn't v2 a subset of v1?

Unfortunately we need both. Jenkins will complain about v1 because of extra XML attributes like `file` and `line` in XML test case element. GitLab CI can use `file` attribute to link source file to test in their tests report.

- Jenkins is very happy with v2 but very unhappy with v1. Jenkins XML parser with their schema is very strict
- GitLab is happy with both but only with v1 it can link source file to test in their tests report: [GitLab JUnit format specification](https://docs.gitlab.com/ci/testing/unit_test_reports/#junit-xml-format-specification). GitLab CI XML parser is very... forgivable because it is written in Ruby [junit.rb](https://gitlab.com/gitlab-org/gitlab/-/blob/v18.7.0-ee/lib/gitlab/ci/parsers/test/junit.rb?ref_type=tags) without any strict schema
- Lot of other CI environments have similar story :melting_face: 

v2 has lot of things in common with v1 but in the same time some parts are changed/deprecated/removed. It depends how XML parser is strict :thinking: Be strict like Jenkins or be happy with everything like GitLab.

> Then we at least need a way to let users configure what they want.

Yes. I will add `COCOTB_RESULTS_FORMAT` environment variable. And also `COCOTB_RESULTS_WORKSPACE` because some CI environments prefer relative paths and don't work well with absolute paths (GitLab CI :eyes:). I will also rename `COCOTB_ATTACHMENTS` to `COCOTB_RESULTS_ATTACHMENTS` :thinking: Then everything will be covered by `COCOTB_RESULTS_*` environment variables.

tymonx commented:
> I'd argue that's acceptable. IMO, the benefits of always writing the same format (v2) that's universally understood outweighs the downside of having to detect the environment we're running in.

Pytest by default it is using v2 with an option to change it via configuration file like `pyproject.toml` file or command line argument: `pytest --override-ini=junit_family=xunit1`. We can go with that route as well.

Another question is relative paths needed in `file` attributes but also in attachment properties. We can assume current working directory from cocotb runner point of view with an option to override it using the `COCOTB_RESULTS_WORKSPACE` environment variable or something similar if someone would try to `cd` or invoke `pytest` from different location in their CI pipelines/jobs. Please note that current working directory of simulator is changed to test directory != project/workspace directory.

We could drop auto-detection because there are 10-20 different CI environments to support. To high maintenance burn. Override format/workspace could be done via environment variables (later from configuration files and `pytest` command line arguments if https://github.com/cocotb/cocotb/pull/5090 will be merged).

ktbarrett commented:
That is to say I think defaulting to xunit2 and opportunistically adding the xunit1 functionality is great. And a switch to decide to add that functionality in cases where detection doesn't work correctly is great. But we need a well defined base case that works for everyone we currently know about and then just flags for additional functionality on top of that.

And that entails us detailing what that common format is. Hopefully that's something simple like "it's just xunit2, see here, and these properties: ...".

I don't want us to have to explicitly support every CI environment.

tymonx commented:
Timeline is strange because it is highly probably that pytest with proper built-in support for JUnit XML will land with this PR in the same time or relative close. Support for v2 without any kind of auto-detection should be enough. And we should recommend pytest.

This PR will be also a good playground to test and gather feedback for file attachments (waveform). This is something not natively supported in vanilia pytest. It can be done in few lines by providing custom plugin or using an external existing one.

tymonx commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -140,7 +140,7 @@ def add_testcase(
             classname:   Path to module (using the ``.`` dot as separator) where test is located.
             file:        Path to file with test.
             line:        Line number of test in file.
```
@ktbarrett Fixed in this commit 38034f5f7ec91f5b12dc5b2e3d0a2fd057d3369a for all docstring arguments.

tymonx commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,398 @@
+        """Create new instance of xUnit reporter.
+
+        Args:
+            name:        Name of xUnit reporter.
```
@cmarqu Fixed in this commit 38034f5f7ec91f5b12dc5b2e3d0a2fd057d3369a for all docstring arguments.

tymonx commented:
To validate XML against schema, first download it:

```plaintext
curl -sfLo junit.xsd https://raw.githubusercontent.com/jenkinsci/xunit-plugin/refs/tags/xunit-2.4.0/src/main/resources/org/jenkinsci/plugins/xunit/types/model/xsd/junit-10.xsd
```

Then invoke the `xmllint` command line with downloaded schema and generated XML:

```plaintext
xmllint --schema junit.xsd junit.xml
```

Output:

```plaintext
junit.xml:5: element properties: Schemas validity error : Element 'properties': This element is not expected.
```

With Jenkins xUnit schema version 2 it complains only about `properties` XML element in `testcase` XML element :roll_eyes: Without it, it is not possible to expose custom data like simulation time duration per test. Officially it is supported in `testsuite` XML element. Other schemas are allow it.

tymonx commented:
The Jenkins [xunit-plugin](https://github.com/jenkinsci/xunit-plugin) is also supporting `properties` in `testcase` XML element. They have some test samples: [googletest/testcase4/input.xml](https://github.com/jenkinsci/xunit-plugin/blob/8c515de2804b3fd591a3b0f1ff8908a1b1a41887/src/test/resources/org/jenkinsci/plugins/xunit/types/googletest/testcase4/input.xml#L84). Both plugins are support that.

tymonx commented:
After some digging. This is from the [jest-unit](https://github.com/jest-community/jest-junit?tab=readme-ov-file#adding-custom-testcase-properties) project:

> WARNING: Properties for testcases is not following standard JUnit XML schema. However, other consumers may support properties for testcases like [DataDog metadata through <property> elements](https://docs.datadoghq.com/continuous_integration/tests/junit_upload/?tab=jenkins#providing-metadata-through-property-elements)

And this only confirm my suspicion. `properties` in `testcase` XML element are not officially allowed in [Jenkins JUnit schema](https://github.com/jenkinsci/xunit-plugin/blob/xunit-2.4.0/src/main/resources/org/jenkinsci/plugins/xunit/types/model/xsd/junit-10.xsd) but lot of tools are supporting them anyway including Jenkins plugins like [xunit-plugin](https://github.com/jenkinsci/xunit-plugin) or [junit-plugin](https://github.com/jenkinsci/junit-plugin) :roll_eyes: 

Case solved.

tymonx commented:
The only valid response to this:
<img width="666" height="500" alt="image" src="https://github.com/user-attachments/assets/8760e3b1-3249-4cb3-a9e5-3d7f14bcdfbf" />

ktbarrett reviewed (commented):
I guess speaking of the RANDOM_SEED issues, does the pytest regression manager do the same thing where the seed is set to a unique, but repeatable value on a per-test basis so not all tests in a regression are running with the same seed and the random state isn't bleeding from test to test?

tymonx commented on docs/source/library_reference.rst:
```diff
@@ -287,6 +287,19 @@ Discovering Tests
+
+    .. versionadded:: 2.1
+
+.. envvar:: COCOTB_ATTACHMENTS
```
I don't have any specific preferences.

- So `COCOTB_RESULTS_ATTACHMENTS` :thinking: I'm fine with that.
- So `COCOTB_RESULTS_RELATIVE_TO` :thinking: I'm also fine with that. `COCOTB_RESULTS_PATHS_RELATIVE_TO` could be too long. 
- This will be aligned with existing `COCOTB_RESULTS_FILE` environment variable (the same prefix `COCOTB_RESULTS_*`)

tymonx commented on src/cocotb/regression.py:
```diff
@@ -630,17 +644,17 @@ def _record_test_init_failed(self) -> None:
+            # Reason
+            status="error",
+            reason="Test initialization failed",
+            system_err=f"Test failed with COCOTB_RANDOM_SEED={cocotb.RANDOM_SEED}",
```
That string with `cocotb.RANDOM_SEED` is related with this reported issue https://github.com/cocotb/cocotb/issues/5185

tymonx commented:
> I guess speaking of the RANDOM_SEED issues, does the pytest regression manager do the same thing where the seed is set to a unique, but repeatable value on a per-test basis so not all tests in a regression are running with the same seed and the random state isn't bleeding from test to test?

@ktbarrett I have copied from built-in regression manager but from version 1.9.1 or 2.0.0 dev. Current [src/cocotb/regression.py](https://github.com/cocotb/cocotb/blob/fe7d22e9185d8c37773c0699f0dc785f2a254f52/src/cocotb/regression.py#L385) has:

```python
# seed random number generator based on test module, name, and COCOTB_RANDOM_SEED
hasher = hashlib.sha1()
hasher.update(self._test.fullname.encode())
test_seed = self._regression_seed + int(hasher.hexdigest(), 16)

cocotb.RANDOM_SEED = test_seed
self._random_state = random.getstate()
random.seed(test_seed)
```

But pytest plugin [src/cocotb_tools/pytest/regression.py](https://github.com/cocotb/cocotb/blob/fe7d22e9185d8c37773c0699f0dc785f2a254f52/src/cocotb_tools/pytest/regression.py#L416):

```python
# seed random number generator based on test module, name, and COCOTB_RANDOM_SEED
hasher = hashlib.sha1()
hasher.update(item.nodeid.encode())
seed = cocotb.RANDOM_SEED + int(hasher.hexdigest(), 16)
random.seed(seed)
```

I will align that part of code.

ktbarrett reviewed (approved):
Looks good to me. But we definitely want other maintainers to have a look.

imphil commented:
I started a review, but will need a couple more days to finish it. I'll get back to you.

ktbarrett commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+Status = Literal["passed", "failed", "skipped", "error"]
+"""Status of test case.
+
+- ``passed`` - Test passed. Default status.
```
```suggestion
- ``passed`` - Test passed.
```

ktbarrett commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+        self._testsuites: dict[str, TestSuite] = {}
+
+        # Common properties that will be added to all created test cases
+        self._properties: dict[str, Any] = dict(kwargs)
```
```suggestion
        self._properties: dict[str, Any] = dict(properties)
```

ktbarrett commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+        system_err: str = "",
+        status: Status = "passed",
+        reason: str | BaseException | None = None,
+        **kwargs: Any,
```
Same here with explicitly separating properties to prevent namespace collisions.
```suggestion
        extra_properties: Mapping[str, Any],
```

ktbarrett commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+            system_err: Text that will be included in the ``system-err`` element.
+            status: Status of test case.
+            reason: Reason of failed or skipped test case.
+            kwargs: Additional test case properties.
```
```suggestion
            extra_properties: Additional test case properties.
```

ktbarrett commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+            system_err += os.linesep
+
+        properties: dict[str, Any] = self._properties.copy()
+        properties.update(kwargs)
```
```suggestion
        properties.update(extra_properties)
```

ktbarrett replied:
```suggestion
        properties: dict[str, Any] = self._default_properties.copy()
        properties.update(extra_properties)
```

ktbarrett commented on docs/source/newsfragments/5219.removal.1.rst:
```diff
@@ -0,0 +1 @@
+Removed support for undocumented ``COCOTB_RESULT_TESTSUITE`` environment variable. Name of test suite will be based on name of Python module.
```
```suggestion
Removed support for undocumented ``COCOTB_RESULT_TESTSUITE`` environment variable. Name of test suite will be based on name of the cocotb test module.
```

ktbarrett commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+
+Simulation log and waveform files will be attached to generated JUnit XML file by using :ref:`junit-attributes-property` and
+:ref:`junit-attributes-system-out` XML elements. List of files that should be attached can be defined by using
+the :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable.
```
FFR we split lines based on sentences (or if the line gets really long). This reduces diffs in the future.
```suggestion
Simulation log and waveform files will be attached to generated JUnit XML file by using :ref:`junit-attributes-property` and :ref:`junit-attributes-system-out` XML elements.
The :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable is used to define the list of attachments.
```

ktbarrett commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+:ref:`junit-attributes-system-out` XML elements. List of files that should be attached can be defined by using
+the :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable.
+
+.. note::
```
```suggestion
pytest Plugin
-------------
```

ktbarrett commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+
+    COCOTB_RESULTS_RELATIVE_TO="${CI_PROJECT_DIR:-$(pwd)}" pytest|make ...
+
+.. note::
```
```suggestion
pytest Plugin
-------------
```

ktbarrett commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+Attributes
+==========
+
+Only subset of XML elements and attributes from `Jenkins xUnit schema version 2`_ are supported.
```
```suggestion
Only a subset of XML elements and attributes from `Jenkins xUnit schema version 2`_ are supported.
```

ktbarrett reviewed (commented):
The approach seems good. Mostly nits IMO.

The only thing I see with the documentation conceptually is that there is a "how-to" section followed by a "reference" section. See https://diataxis.fr/ for the system we are trying to follow.

Maybe we split up the first couple sections about attachments and paths and the sections after that. The first couple sections can stay in the index where they are, and the rest can go into a `junit_reference.rst` and put in the reference section under the `library_reference`. Then maybe in the how-to docs in the preamble add a blurb along the lines of "If you need to know what set of attributes are supported by the cocotb XML result file, see [link to reference manual page]".

ktbarrett reviewed (approved):
I'll ping the other maintainers as a courtesy. If they have no response in a couple days I'll merge. Thanks again for threading the needle on the weird multiple conflicting result file formats.

tymonx commented on src/cocotb/_xunit_reporter.py:
```diff
@@ -3,16 +3,332 @@
+        system_out: str = "",
+        system_err: str = "",
+        status: Status = "passed",
+        reason: str | BaseException | None = None,
```
https://github.com/tymonx/cocotb/blob/4d8501dc7536ea3e4d2bf9445774dfcbef674350/src/cocotb/_xunit_reporter.py#L104-L114

@ktbarrett `classname` and `status` arguments are now mandatory. I kept `reason` as optional with default to `None`. `reason=None` with `status="passed"` looked strange. The same story with the `time` argument. It looked strange for skipped tests and failed during initialization. But I can change that also.

For 100% I would kept `system_out`, `system_err` and `extra_properties` as optional (with default empty values).

tymonx commented:
> Maybe we split up the first couple sections about attachments and paths and the sections after that. The first couple sections can stay in the index where they are, and the rest can go into a junit_reference.rst and put in the reference section under the library_reference.

I have split documentation about JUnit Tests Report to separate `junit.rst` and `junit_reference.rst` files as suggested. After I applied all suggestions. I think it should be OK now :thinking: 

>  https://diataxis.fr/

Thanks for that! I didn't know that there is a term for that :slightly_smiling_face:

imphil reviewed (commented):
Thanks for your work on researching, implementing and documenting the output format. Much appreciated!

It took me a while to understand two things:

* You're documenting a file format that's produced by two very different tools. cocotb, if the Makefiles or cocotb runner are used, or pytest (when the pytest support is used and pytest is configured in a certain way). This needs to be spelled out much more clearly.

* Attachments are not actually attached, as in, included with the test reports. They are just references to files, links, or whatever you might call them. So you're not "attaching the file", you're "attaching a reference to the file". That, again, need to be spelled out more explicitly.

Please try to make those points more clear to the user.

On type annotations: please skip them if they don't provide any value, i.e., if type checkers can infer what you mean. They are just visual clutter. 

To me, your PR was rather hard to review, which slows the review process down significantly (I can squeeze in 10 minutes between other things easily, but if I have to block out multiple hours from a day that's not so easy). 

Please: 
* Ensure your PR solves *one* problem. Describe this problem in the PR description.
* As soon as you're tempted to write a bulleted list, you probably did more than one thing. Either split the change into more than one PR, or split the PR into a series of *atomic* commits, each with a descriptive commit message. For example, support for attachments is at least worth a separate commit.
* Give yourself a bit of cooldown time after addressing review comments: read the comments, make your changes, do a self-review, and after you're happy, update the PR and let us know what changed (and why). It's easier to read a block of code and see what changed than being in a constant stream of chat-like updates. 

(There are most likely some review comments that are outdated and I can't delete any more, since your latest changes collided with an outstanding review. I'll try to close those after the review is submitted.)

tymonx commented on src/cocotb_tools/pytest/_regression.py:
```diff
@@ -693,6 +706,22 @@ def pytest_runtest_makereport(
+            # pytest --override-ini=junit_logging=system-out|all ...
+            report.sections.append(("Captured stdout", f"[[ATTACHMENT|{attachment}]]"))
+
+        # FIXME: This is not working yet, it requires a fix in pytest internal plugin: junitxml
```
I will re-phrase that comment. Pytest `TestReport` objects are propagating information about tests to all reporters. This information will be properly logged to console, log file and any pytest plugins that are processing test reports.

But the built-in pytest `junitxml` plugin has a specific "issue" https://github.com/cocotb/cocotb/issues/5226

This part of code is valid. Comment is too misleading.

tymonx commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+Only few XML elements and attributes are used from above XML schema when generating XML tests report
+that are commonly supported in other available JUnit/xUnit XML schemas.
+
+This allow to support wide range of different CI environments like GitHub Actions, GitLab CI, Jenkins CI and others.
```
> Does GHA even parse the junit xml file itself? Or only with Action?

@imphil Unfortunately not. You must use one of existing actions like [junit-report-action](https://github.com/marketplace/actions/junit-report-action). I should add this information as well :thinking: I could add an example to these three CI environments. But generally there are dozens of CI environments.

tymonx commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,374 @@
+
+List of supported properties:
+
+* ``cocotb`` - mark ``testcase`` as cocotb test. If defined, default value is ``True``.
```
@imphil I can re-phrase that or just remove it. If you have multiple JUnit XML tests reports, generated from C/C++ software, Python and Cocotb, it is a good idea to have some kind of identifier after merging all tests reports into single tests report (this is often happening automatically on CI side) to distinguish them.

tymonx commented:
@ktbarrett Yes. I will have some time this week (at the end). I will try to synchronize my PR today, test if still everything works and apply all suggestions. Mostly PR should be ready around this weekend.

ktbarrett commented:
Christ Windows you suck.

EDIT: Determined the issue was XML not Windows, which sucks almost as bad as Windows. `\r` is not a legal character in XML, so for whatever reason, `ElementTree.write` converts `\r` into `\n`, leading to `\n\n`.

cmarqu commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,97 @@
+pytest Plugin
+-------------
+
+Set the ``junit_logging`` option to ``system-out`` or ``all``.
```
...to do what? Is this a prerequisite for being able to attach files?

ktbarrett replied:
I actually have no idea. I thought the same thing.

ktbarrett replied:
```suggestion
Setting the ``junit_logging`` pytest INI config option to ``system-out`` or ``all`` is required for attachments to be added when using the pytest plugin.
```

cmarqu commented on docs/source/junit.rst:
```diff
@@ -0,0 +1,97 @@
+Outputting Relative Paths
+=========================
+
+The :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable can be used to convert all absolute paths reported in the generated XML file to relative ones including XML attributes, properties and text.
```
```suggestion
The :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable can be used to convert all absolute paths reported in the generated XML file to relative ones, including XML attributes, properties and text.
```
?

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+
+This document describes the format of the JUnit XML test report files generated by cocotb.
+
+This formats aims to support wide range of different CI environments like:
```
```suggestion
This formats aims to support a wide range of different CI environments like:
```

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+
+This formats aims to support wide range of different CI environments like:
+
+* GitHub Actions (with the help of junit-report-action or similar actions)
```
```suggestion
* GitHub Actions (with the help of ``junit-report-action`` or similar actions)
```

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+* Jenkins CI
+
+The result XML files are compatible with `Jenkins xUnit schema version 2`_
+and with JUnit XML test report files from pytest when used with ``junit_family=xunit2`` pytest option.
```
```suggestion
and with JUnit XML test report files from pytest when used with the ``junit_family=xunit2`` pytest option.
```

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+* ``tests`` - total number of tests reported in this testsuite.
+* ``time`` - total wall time of all tests in this testsuite in seconds.
+* ``timestamp`` - date and time of when the testsuite run was executed (in the ISO 8601 format with the UTC timezone).
+* ``hostname`` - name of machine where the testsuite run was executed.
```
```suggestion
* ``hostname`` - name of the machine where the testsuite run was executed.
```

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+
+.. note::
+
+    When generating JUnit XML file from pytest, it is possible to include ``file`` and ``line`` as attributes in ``testcase`` elements by setting the pytest ``junit_family`` to ``xunit1``:
```
```suggestion
    When generating a JUnit XML file from pytest, it is possible to include ``file`` and ``line`` as attributes in ``testcase`` elements by setting the pytest ``junit_family`` to ``xunit1``:
```

cmarqu commented on docs/source/junit_reference.rst:
```diff
@@ -0,0 +1,329 @@
+List of supported attributes:
+
+* ``classname`` - Python module where reported test was defined.
+* ``name`` - name of reported test, mostly name of Python function or Python class.
```
```suggestion
* ``classname`` - Python module where the reported test was defined.
* ``name`` - name of the reported test, mostly the name of a Python function or a Python class.
```

cmarqu reviewed (commented):
Not finished with the review, but need to stop here.

ktbarrett committed:
Apply suggestions from code review

Co-authored-by: Colin Marquardt <cmarqu42@gmail.com>
Co-authored-by: Kaleb Barrett <dev.ktbarrett@gmail.com>

imphil reviewed (approved):
I trust Kaleb on this without looking much further. Thanks for finishing this PR!

@ktbarrett I'll let you squash/merge this PR when you're ready.
